cpp/src/arrow/array.h

```diff
@@ -367,8 +367,8 @@ class ARROW_EXPORT NullArray : public FlatArray {
  private:
   inline void SetData(const std::shared_ptr<ArrayData>& data) {
     null_bitmap_data_ = NULLPTR;
-    data->null_count = data->length;
-    data_ = data;
+    data_ = data->Copy();
+    data_->null_count = data->length;
```

Comment:
I do not think copying is correct. `ArrayData` is an internal API; copying should be addressed at a higher level

Reply:
I'll revert it, then.